aws/lambda/pytorch-auto-revert/pytorch_auto_revert/autorevert_checker.py

```diff
@@ -74,13 +84,24 @@ def __init__(
         self._workflow_commits_cache: Dict[str, List[CommitJobs]] = {}
         self._commit_history = None
         self._ignore_classification_rules = ignore_classification_rules or set()
+        # Controls whether queries target restarted runs only (workflow_dispatch/tagged trunk/<sha>)
+        self._use_restarted_runs_only = False
 
     def get_workflow_commits(self, workflow_name: str) -> List[CommitJobs]:
         """Get workflow commits for a specific workflow, fetching if needed. From newer to older"""
         if workflow_name not in self._workflow_commits_cache:
             self._fetch_workflow_data()
         return self._workflow_commits_cache.get(workflow_name, [])
 
+    def get_workflow_commit_by_sha(
+        self, workflow_name: str, sha: str
+    ) -> Optional[CommitJobs]:
+        """Return CommitJobs for a workflow and head_sha if present in cache."""
+        for cj in self.get_workflow_commits(workflow_name):
```

Comment:
this loops happens inside a loop, I suspect that when running it for a long period like 2 years for evaluation this would be a significant bottleneck in the code due the quadratic nature.

Maybe the search here could leverage a map...?

Reply:
trie would be even better for prefix search, but I don't think that would be a bottleneck